File: PJ_2019/src/Polarimeter/frmColorTable.vb
Public Class frmColorTable

    Private Sub frmColorTable_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        LoadColorTable()
    End Sub

    Private Sub LoadColorTable()
        lvColorTable.Items.Clear()
        Dim LVI As ListViewItem
        LVI = New ListViewItem
        LVI.UseItemStyleForSubItems = False
        LVI.Text = "Reference"
        LVI.SubItems.Add("")
        LVI.SubItems.Add(frmMain.ReferenceColor.ToString)
        LVI.SubItems(1).BackColor = frmMain.ReferenceColor
        lvColorTable.Items.Add(LVI)
        For i As Integer = 0 To 19
            LVI = New ListViewItem
            LVI.UseItemStyleForSubItems = False
            LVI.Text = "Sample " & (i + 1).ToString
            LVI.SubItems.Add("")
            LVI.SubItems.Add(frmMain.ColorTable(i).ToString)
            LVI.SubItems(1).BackColor = frmMain.ColorTable(i)
            lvColorTable.Items.Add(LVI)
        Next
    End Sub

    Private Sub lvColorTable_MouseDoubleClick(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles lvColorTable.MouseDoubleClick
        Try
            Dim i As Integer = lvColorTable.SelectedIndices(0)
            Dim dlg As New ColorDialog
            dlg.Color = lvColorTable.SelectedItems(0).SubItems(1).BackColor
            dlg.ShowDialog()
            If i = 0 Then
                frmMain.ReferenceColor = dlg.Color
            Else
                frmMain.ColorTable(i - 1) = dlg.Color
            End If
            lvColorTable.SelectedItems(0).SubItems(1).BackColor = dlg.Color
            lvColorTable.SelectedItems(0).SubItems(2).Text = dlg.Color.ToString
            Select Case i
                Case 0
                    My.Settings.ReferenceColor = dlg.Color
                Case 1
                    My.Settings.Color1 = dlg.Color
                Case 2
                    My.Settings.Color2 = dlg.Color
                Case 3
                    My.Settings.Color3 = dlg.Color
                Case 4
                    My.Settings.Color4 = dlg.Color
                Case 5
                    My.Settings.Color5 = dlg.Color
                Case 6
                    My.Settings.Color6 = dlg.Color
                Case 7
                    My.Settings.Color7 = dlg.Color
                Case 8
                    My.Settings.Color8 = dlg.Color
                Case 9
                    My.Settings.Color9 = dlg.Color
                Case 10
                    My.Settings.Color10 = dlg.Color
                Case 11
                    My.Settings.Color11 = dlg.Color
                Case 12
                    My.Settings.Color12 = dlg.Color
                Case 13
                    My.Settings.Color13 = dlg.Color
                Case 14
                    My.Settings.Color14 = dlg.Color
                Case 15
                    My.Settings.Color15 = dlg.Color
                Case 16
                    My.Settings.Color16 = dlg.Color
                Case 17
                    My.Settings.Color17 = dlg.Color
                Case 18
                    My.Settings.Color18 = dlg.Color
                Case 19
                    My.Settings.Color19 = dlg.Color
                Case 20
                    My.Settings.Color20 = dlg.Color
            End Select
            My.Settings.Save()
            frmMain.ApplyColorTableToSamples()
        Catch ex As Exception
            'do nothing
        End Try
    End Sub

    Private Sub btnClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClose.Click
        Close()
    End Sub

End Class

File: Polarimeter2019/frmColorTable.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Polarimeter2019
{
    public partial class frmColorTable : Form
    {
        frmMain f1;

        public frmColorTable(frmMain f)
        {
            InitializeComponent();
            f1 = f;
        }

        private void frmColorTable_Load(object sender, EventArgs e)
        {
            LoadColorTable();
        }

        private void LoadColorTable()
        {
            lvColorTable.Items.Clear();
            ListViewItem LVI = new ListViewItem();
            LVI.UseItemStyleForSubItems = false;
            LVI.Text = "Reference";
            LVI.SubItems.Add("");
            LVI.SubItems.Add(f1.ReferenceColor.ToString());
            LVI.SubItems[1].BackColor = f1.ReferenceColor;
            lvColorTable.Items.Add(LVI);

            for (int i = 0; i <= 19; i++)
            {
                LVI = new ListViewItem();
                LVI.UseItemStyleForSubItems = false;
                LVI.Text = "Sample " + (i + 1).ToString();
                LVI.SubItems.Add("");
                LVI.SubItems[1].BackColor = f1.ColorTable[i];
                lvColorTable.Items.Add(LVI);
            }
        }

        private void lvColorTable_MouseDoubleClick(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            try
            {
                int i = lvColorTable.SelectedIndices[0];
                ColorDialog dlg = new ColorDialog();
                colorDialog1.ShowDialog();
                dlg.Color = lvColorTable.SelectedItems[0].SubItems[1].BackColor;
                dlg.ShowDialog();
                if (i == 0)
                {
                   f1.ReferenceColor = dlg.Color;
                }
                else
                {
                    f1.ColorTable[i - 1] = dlg.Color;
                }

                lvColorTable.SelectedItems[0].SubItems[1].BackColor = dlg.Color;
                lvColorTable.SelectedItems[0].SubItems[2].Text = dlg.Color.ToString();
            switch (i)
            {
                case 0:
                    {
                        Properties.Settings.Default.ReferenceColor = dlg.Color;
                        break;
                    }

                case 1:
                    {
                        Properties.Settings.Default.color1 = dlg.Color;
                        break;
                    }

                case 2:
                    {
                        Properties.Settings.Default.color2 = dlg.Color;
                        break;
                    }

                case 3:
                    {
                        Properties.Settings.Default.color3 = dlg.Color;
                        break;
                    }

                case 4:
                    {
                        Properties.Settings.Default.color4 = dlg.Color;
                        break;
                    }

                case 5:
                    {
                        Properties.Settings.Default.color5 = dlg.Color;
                        break;
                    }

                case 6:
                    {
                        Properties.Settings.Default.color6 = dlg.Color;
                        break;
                    }

                case 7:
                    {
                        Properties.Settings.Default.color7 = dlg.Color;
                        break;
                    }

                case 8:
                    {
                        Properties.Settings.Default.color8 = dlg.Color;
                        break;
                    }

                case 9:
                    {
                        Properties.Settings.Default.color9 = dlg.Color;
                        break;
                    }

                case 10:
                    {
                        Properties.Settings.Default.color10 = dlg.Color;
                        break;
                    }

                case 11:
                    {
                        Properties.Settings.Default.color11 = dlg.Color;
                        break;
                    }

                case 12:
                    {
                        Properties.Settings.Default.color12 = dlg.Color;
                        break;
                    }

                case 13:
                    {
                        Properties.Settings.Default.color13 = dlg.Color;
                        break;
                    }

                case 14:
                    {
                        Properties.Settings.Default.color14 = dlg.Color;
                        break;
                    }

                case 15:
                    {
                        Properties.Settings.Default.color15 = dlg.Color;
                        break;
                    }

                case 16:
                    {
                        Properties.Settings.Default.color16 = dlg.Color;
                        break;
                    }

                case 17:
                    {
                        Properties.Settings.Default.color17 = dlg.Color;
                        break;
                    }

                case 18:
                    {
                        Properties.Settings.Default.color18 = dlg.Color;
                        break;
                    }

                case 19:
                    {
                        Properties.Settings.Default.color19 = dlg.Color;
                        break;
                    }

                case 20:
                    {
                        Properties.Settings.Default.color20 = dlg.Color;
                        break;
                    }
            }
            Properties.Settings.Default.Save();
            f1.ApplyColorTableToSamples();
            }
            catch (Exception ex)
            {

            }
        }

        private void btnOK_Click(object sender, EventArgs e)
        {
            this.DialogResult = DialogResult.OK;
            this.Close();
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void lvColorValue_DoubleClick(object sender, EventArgs e)
        {
            colorDialog1.ShowDialog();
            lvColorTable.SelectedItems[0].SubItems[1].BackColor = colorDialog1.Color;
        }
    }
}
